Mathlib/LinearAlgebra/OnSup.lean

```diff
@@ -0,0 +1,172 @@
+  quotKerEquivOfSurjective _ (by
+    rw [← range_eq_top, eq_top_iff]
+    rintro ⟨x, hx⟩ _
+    obtain ⟨y, hy, z, hz, rfl⟩ := mem_sup.mp hx
+    use ⟨⟨y, hy⟩, ⟨z, hz⟩⟩
+    simp [coprod_apply, ← Subtype.coe_inj, coe_add, coe_inclusion])
+
+namespace LinearMap
+
+variable {N : Type*} [AddCommGroup N] [Module R N] {f : P →ₗ[R] N} {g : Q →ₗ[R] N}
+
+/-- Given two linear maps `f : P →ₗ[R] N` and `g : Q →ₗ[R] N` that agree on `P ⊓ Q`, this is
+the linear map `P ⊔ Q →ₗ[R] N` that simultaneously extends `f` and `g`. -/
+noncomputable def onSup
+    -- (h : ∀ x (hP : x ∈ P) (hQ : x ∈ Q), f ⟨x, hP⟩ = g ⟨x, hQ⟩) :
+    (h : f ∘ₗ inclusion inf_le_left = g ∘ₗ inclusion inf_le_right) :
+    ↥(P ⊔ Q) →ₗ[R] N := by
+  apply comp ((ker _).liftQ (f.coprod g) ?_) quotientCoprodAddEquiv.symm.toLinearMap
+  rintro ⟨⟨x, hx⟩, ⟨y, hy⟩⟩ hxy
+  simp only [mem_ker, coprod_apply, add_eq_zero_iff_eq_neg, ← Subtype.coe_inj, coe_inclusion,
+    NegMemClass.coe_neg] at hxy
+  simp only [mem_ker, coprod_apply, add_eq_zero_iff_eq_neg, ← map_neg]
+  have hx' : x ∈ P ⊓ Q := ⟨hx, by simp [hxy, hy]⟩
+  rw [show ⟨x, hx⟩ = inclusion inf_le_left ⟨x , hx'⟩ by simp [← Subtype.coe_inj]]
+  rw [show -⟨y, hy⟩ = inclusion inf_le_right ⟨x, hx'⟩ by simp [← Subtype.coe_inj, hxy]]
+  simp only [← LinearMap.comp_apply, h]
+
+theorem onSup_apply_left
+    (h : f ∘ₗ inclusion inf_le_left = g ∘ₗ inclusion inf_le_right)
+    {x : M} (hx : x ∈ P) : onSup h ⟨x, le_sup_left (b := Q) hx⟩ = f ⟨x, hx⟩ := by
+  have h : (quotientCoprodAddEquiv (P := P) (Q := Q)).symm ⟨x, le_sup_left (b := Q) hx⟩ =
+      Submodule.Quotient.mk ⟨⟨x, hx⟩, (0 : Q)⟩ := by
+    simp [symm_apply_eq, quotientCoprodAddEquiv, quotKerEquivOfSurjective, inclusion_apply]
+  simp only [add_eq_sup, onSup, comp_apply] at h ⊢
+  simp [h, liftQ_apply]
+
+theorem onSup_apply_right
+    (h : f ∘ₗ inclusion inf_le_left = g ∘ₗ inclusion inf_le_right)
+    {x} (hx : x ∈ Q) : onSup h ⟨x, le_sup_right (a := P) hx⟩ = g ⟨x, hx⟩ := by
+  have h : (quotientCoprodAddEquiv (P := P) (Q := Q)).symm ⟨x, le_sup_right (a := P) hx⟩ =
+      Submodule.Quotient.mk ⟨(0 : P), ⟨x, hx⟩⟩ := by
+    simp [symm_apply_eq, quotientCoprodAddEquiv, quotKerEquivOfSurjective, inclusion_apply]
+  simp only [add_eq_sup, onSup, comp_apply] at h ⊢
+  simp [h, liftQ_apply]
+
+theorem onSup_apply
+    (h : f ∘ₗ inclusion inf_le_left = g ∘ₗ inclusion inf_le_right)
+    {x y} (hx : x ∈ P) (hy : y ∈ Q) :
+    onSup h ⟨x + y, add_mem_sup hx hy⟩ = f ⟨x, hx⟩ + g ⟨y, hy⟩ := by
+  simp [← onSup_apply_left h hx, ← onSup_apply_right h hy, ← map_add]
+
+theorem onSup_comp_left (h : f ∘ₗ inclusion inf_le_left = g ∘ₗ inclusion inf_le_right) :
+    (onSup h).comp (inclusion le_sup_left) = f := by
+  ext ⟨x, hx⟩
+  simp [← onSup_apply_left h, inclusion_apply]
+
+theorem onSup_comp_right (h : f ∘ₗ inclusion inf_le_left = g ∘ₗ inclusion inf_le_right) :
+    (onSup h).comp (inclusion le_sup_right) = g := by
+  ext ⟨x, hx⟩
+  simp [← onSup_apply_right h, inclusion_apply]
+
+@[ext]
```

Comment:
This probably needs `high` to apply below.

Reply:
it doesn't seem.